scio-core/src/main/scala-3/com/spotify/scio/schemas/To.scala

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2019 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.schemas
+
+import org.apache.beam.sdk.schemas.{SchemaCoder, Schema => BSchema}
+import BSchema.{ FieldType => BFieldType }
+
+import scala.compiletime._
+import scala.deriving._
+import scala.quoted._
+import scala.reflect.ClassTag
+
+object ToMacro {
+
+  def safeImpl[I: scala.quoted.Type, O: scala.quoted.Type](
+    iSchema: Expr[Schema[I]],
+    oSchema: Expr[Schema[O]]
+  )(using Quotes): Expr[To[I, O]] = {
+    import scala.quoted.quotes.reflect.{report, TypeRepr}
+
+    (interpret[I] , interpret[O]) match {
+      case (None, None) => report.throwError(
+        s"""
+        |Could not interpret input schema:
+        |  ${iSchema.show}
+        |Could not interpret output schema:
+        |  ${oSchema.show}
+        |""".stripMargin
+      )
+      case (None, _) => report.throwError("Could not interpret input schema: " + iSchema.show)
+      case (_, None) => report.throwError("Could not interpret output schema: " + oSchema.show)
+      case (Some(sIn), Some(sOut)) =>
+        val schemaOut: BSchema = SchemaMaterializer.fieldType(sOut).getRowSchema()
+        val schemaIn: BSchema = SchemaMaterializer.fieldType(sIn).getRowSchema()
+        val classTagOpt = Expr.summon[ClassTag[O]]
+        if (classTagOpt.isEmpty) {
+          report.throwError(s"Could not summon Expr[ClassTag[${TypeRepr.of[O].show}]]")
+        }
+        val classTag = classTagOpt.get
+        To.checkCompatibility(schemaIn, schemaOut)('{ To.unchecked[I, O](using $iSchema, $oSchema, $classTag) })
+          .fold(message => report.throwError(message), identity)
+    }
+  }
+
+  private def sequence[T](ls: List[Option[T]]): Option[List[T]] =
+    if ls.exists(_.isEmpty) then None
+    else Some(ls.collect { case Some(x) => x })
+
+  private def interpret[T: scala.quoted.Type](using Quotes): Option[Schema[T]] = 
+    Type.of[T] match {
```

Comment:
(copying my comment here since I originally made it on your fork)

Unless I'm missing something this means that the possible set of supported schema types is effectively limited to those already defined in scio. 

If a user adds a `given` instance of `Schema` for type that isn't supported (say for example a Java class), the derivation will simply ignore it. This could be a problem for aliaswa support in `Schema` but probably something we can live with. We don't really expect users to define their own instances of `Schema`.

I'd be curious to see how the implementation looks like for Java beans and Avro's `SpecificRecord`.